src/compiler/parser.ts

```diff
@@ -2649,6 +2694,26 @@ namespace ts {
             return token() === SyntaxKind.CloseParenToken || isStartOfParameter() || isStartOfType();
         }
 
+        function parseJSDocPostfixTypeOrHigher(): TypeNode {

```

Comment:
Seems like solving 1 would require ditching 2. If we want to be able to parse `foo?[]?[]`, we would need to handle postfix types in a uniform way.